src/Microsoft.OData.Core/ODataMetadataOutputContext.cs

```diff
@@ -127,8 +179,18 @@ protected override void Dispose(bool disposing)
             {
                 if (this.xmlWriter != null)
                 {
-                    // XmlWriter.Flush will call the underlying Stream.Flush.
-                    this.xmlWriter.Flush();
+                    // In the async case the underlying stream is the async buffered stream, so we have to flush that explicitly.
+                    if (this.asynchronousOutputStream != null)
+                    {
+                        this.xmlWriter.FlushAsync();

```

Comment:
We don't wait for the async operation to complete?

Reply:
I checked with others codes in this project. For example here: [https://github.com/OData/odata.net/blob/master/src/Microsoft.OData.Core/Json/ODataJsonTextWriter.cs#L298](https://github.com/OData/odata.net/blob/master/src/Microsoft.OData.Core/Json/ODataJsonTextWriter.cs#L298)

I'd invite @Michael Pizzo to share us the reason?


---
In reply to: [556556718](https://github.com/OData/odata.net/pull/1966#discussion_r556556718) [](ancestors = 556556718)

Reply:
The difference is, in ODataJsonTextWriter we are returning the task from calling FlushAsync() on the internal textwriter, so the caller can wait on that result. In this case, though, we are continuing to call additional methods that may then run concurrently, potentially leading to threading issues.

---
In reply to: [560398552](https://github.com/OData/odata.net/pull/1966#discussion_r560398552) [](ancestors = 560398552,556556718)